src/std_extensions/collections.rs

```diff
@@ -174,7 +170,12 @@ mod test {
 
     #[test]
     fn test_list_ops() {
-        let reg = &[EXTENSION.to_owned()].into();
+        let reg = &[
+            EXTENSION.to_owned(),
+            PRELUDE.to_owned(),
+            float_types::extension(),
```

Comment:
Why are those two extra extensions required now?

Reply:
Haha. Subtle. Because with the new "binary compute_signature can return a type-scheme", after executing the binary as we did before, we then do call `substitute` (with no args) on the type scheme (which has no type params). This traverses the type applying the identity substitution, and uses the extension registry to recompute TypeBounds of everything - even e.g. `USIZE`.

So, we could skip all of that - if the type scheme we get back from compute_signature has no type-params (and there are no typeargs), we could call `clone()` rather than `substitute`. This might be a performance advantage, but asymptotically it'd still be O(size of Type tree)...

Moreover, I kind of feel this *should* error if those extensions aren't present; when we convert the binary function to a type scheme, then it'll have to do a non-trivial substitution, and we'll almost certainly need these extensions then.

I mean, I guess we could "optimize" more substitutions into clones, i.e. for arbitrary bits of the type tree rather than just the whole tree, e.g. for `CustomType`s with no args - but those are close enough to leaves anyway, so I think the saving could well be even smaller.

Thoughts welcome....

Reply:
That seems a bit of a nonsequitur but that doesn't matter - however, I'm wondering what you mean. Do you have any (more concrete) examples? I prefer not to add a Type variant that's "a monomorphic FunctionType" because, sitting alongside the variant that's a PolyFuncType, it'd be strictly less general.